### ISSUE TITLE
Idea: Support power-profiles-daemon API

### ISSUE THREAD

**Issue by ryanabx** [linrunner/TLP#720]
**Is your feature request related to a problem? Please describe.**

The problem I face is wanting to have the power savings of tlp with the nice interface that DEs provide with power-profiles-daemon.

**Describe the solution you'd like**

An API could be implemented that takes calls to change the power profile in ppd and changes to a relevant profile in tlp instead.
Since ppd is already recommended to be uninstalled, this would be non-intrusive to a user using tlp, and would allow tlp to hook into the UI components created by GNOME and KDE to control power profiles.

**Describe alternatives you've considered**

There is a project called TuneD which provides custom configuration profiles similar to tlp, and is implementing the power-profiles-daemon API due to a proposal to replace it with ppd in Fedora, see: 
https://discussion.fedoraproject.org/t/f40-change-proposal-tuned-replaces-power-profiles-daemon-self-contained/94995
Also see relevant PR: https://github.com/redhat-performance/tuned/pull/579

**Comment by linrunner** [comment]
Hi, thanks for your suggestion. 

However, there is a major obstacle: TLP is _not_ implemented as a _daemon_ [1], but in my opinion such a daemon would be required for a DBUS API. 

[1] https://linrunner.de/tlp/developers/architecture.html

**Comment by ryanabx** [comment]
> Hi, thanks for your suggestion.
> 
> However, there is a major obstacle: TLP is _not_ implemented as a _daemon_ [1], but in my opinion such a daemon would be required for a DBUS API.
> 
> [1] https://linrunner.de/tlp/developers/architecture.html

Thank you for your response! I think you're right that it's probable that a daemon would be required, since that's how ppd is currently implemented. As such it's probably out of scope for this project. I appreciate your work with tlp by the way!

**Comment by tbodt** [comment]
Indeed, a daemon of some kind is required to implement the API. But it would make sense to implement as a separate, mostly independent, systemd service, to be enabled only if needed.

**Comment by tbodt** [comment]
@linrunner Are you open to adding a python script to the project? If not...well, bash is one of the few languages you can't write a dbus service in, but I can rewrite it in pretty much any other language.

**Comment by tbodt** [comment]
Re tracking all system state: Is it not enough to watch for changes to /run/tlp/last_pwr?

Re root privilege: Yeah, seems like it has to become a security boundary. But it's not different from the way power-profiles-daemon itself operates, and if they have it working TLP can too.

Re three power profiles, the PPD APi is able to dynamically report that there are only two profiles. There is also the option of repurposing "balanced" as automatic selection.

**Comment by linrunner** [comment]
I'm currently quite busy with the 1.8 beta test. I'll get back to you as soon as I have time for a proper answer.

**Comment by linrunner** [comment]
@tbodt sorry for the long delay. 

I would say start with your suggestion. Too many users want it. Here's another one: https://github.com/linrunner/TLP/issues/787

However, I would like to start it separately from the main project in its own repository, working title **TLP-PD**. I hope that's OK for you.

I assume that we need three profiles. I'll think about that separately and start a user survey to design it.

**Comment by tbodt** [comment]
Separate repo does make sense. In the end it likely should become a separate package in distros anyway.

Are you planning to do it, or would you like me to try? I have a bunch of half finished python code which I haven't worked on since my first comment here.

**Comment by linrunner** [comment]
I don't speak Python yet, but I've been wanting to start learning it for years. So you have to start ...

**Comment by tbodt** [comment]
I went and pushed the code I've been writing in the TLP repo so far to a branch over here: https://github.com/tbodt/TLP/commit/fd67ee6d2575fef9ec4de6963558f5c025138337. It's more of a stub, it doesn't talk to TLP yet, but KDE's widget does connect to it at least. TODO move it out to its own repo

**Comment by linrunner** [comment]
@tbodt Is it intentional that **tlp-ppd.policy** is installed twice under its original name and modified by sed each time? Shouldn't this happen in the loop as **$(BUS_NAME).policy**? 

https://github.com/tbodt/TLP/blob/ppd/Makefile#L215

**Comment by tbodt** [comment]
Yes, thanks for pointing that out. But I should also check if it actually has to be installed twice. power-profilers-daemon just installs once.

**Comment by linrunner** [comment]
Uninstall is of .policy, .conf, .service is missing as well.

**Comment by tbodt** [comment]
Updated. It looks like it's fine to just install one polkit file, that's not an external interface and is just a way for tlp-ppd to ask polkit whether the request comes from an interactive user.

**Comment by tbodt** [comment]
Relating to the number of profiles, the protocol is in theory highly flexible and would support any set of profiles with any name, but unfortunately KDE Powerdevil (at least) behaves somewhat strangely when the returned set of profiles is different from the 3 that power-profiles-daemon returns.

**Comment by linrunner** [comment]
@tbodt I would prefer that we call the daemon **tlp-pd** (“TLP Profiles Daemon”) instead of **tlp-ppd**. I'd like to avoid having the competition in the name. Please also rename the repo to **TLP-PD**, otherwise I can't fork it and send you changes ... ;-)

**Comment by linrunner** [comment]
> Relating to the number of profiles, the protocol is in theory highly flexible and would support any set of profiles with any name, but unfortunately KDE Powerdevil (at least) behaves somewhat strangely when the returned set of profiles is different from the 3 that power-profiles-daemon returns.

I think the users would insist on three profiles anyway, simply because they are used to it. We might as well call them “power-saver”, “balanced” and “performance” as ppd does.

My (tricky) task is to figure out how to map this to AC/BAT plus “something” in TLP.

**Comment by tbodt** [comment]
> My (tricky) task is to figure out how to map this to AC/BAT plus “something” in TLP.

When I thought about this last time I imagined that you could change 0/1 function arguments to be AC/BAT, expand those directly into the variable name, and now it's possible to make as many profiles as you want by changing the last word of the settings variable. But talk is cheap.

**Comment by tbodt** [comment]
> [@tbodt](https://github.com/tbodt?rgh-link-date=2025-03-24T14%3A33%3A20.000Z) I would prefer that we call the daemon **tlp-pd** (“TLP Profiles Daemon”) instead of **tlp-ppd**. I'd like to avoid having the competition in the name. Please also rename your repo to **TLP-PD**, otherwise I can't fork it and send you changes ... ;-)

I've renamed the repo and also added you as a committer.

**Comment by linrunner** [comment]
@tbodt GitHub doesn't let me fork your repository, perhaps because it is forked from TLP.

I think it can't be avoided that you set up only your contribution, i.e. what is necessary for TLP-PD as a *new* repository without the whole history of TLP. Delete the TLP-PD repository on GitHub beforehand.

So in the first approach LICENSE (GPL2), COPYING (you as author) and Makefile plus your stuff. 

You can leave the Makefile as it is for now, and I'll strip it down.

**Comment by tbodt** [comment]
I'll try that this evening.

**Comment by tbodt** [comment]
Just got around to it :)

**Comment by linrunner** [comment]
Well. I just sat down trying to understand the **Makefile** and the _BUS_NAME_ replacements there and I don't get it (yet):

1. There is a loop for *BUS_NAME* over *org.freedesktop.UPower.PowerProfiles net.hadess.PowerProfiles* creating two renamed copies each of **tlp-pd.dbus.conf** and **tlp-pd.dbus.service** and also doing the replacements for *@BUS_NAME@*.
2. There is an instance of *@BUS_NAME@* in **tlp-pd.service.in** which is replaced by the _$(INFILES)_ target.
3. **But** **tlp-pd** contains three instances of *@BUS_NAME@* that are replaced nowhere - is this intended?
4. Do we really need **both** interfaces _org.freedesktop.UPower.PowerProfiles_ **and** _net.hadess.PowerProfiles_?

**Comment by tbodt** [comment]
> 3. **But** **tlp-pd** contains three instances of _@BUS_NAME@_ that are replaced nowhere - is this intended?
It is replaced at the bottom in python code before bus.publish.

> 4. Do we really need **both** interfaces _org.freedesktop.UPower.PowerProfiles_ **and** _net.hadess.PowerProfiles_?
Sadly, yes...because the PowerProfiles project changed ownership and changed the interface name along with that. But some clients still use the old name, so they have to keep both working.

**Comment by linrunner** [comment]
@tbodt I have been thinking quite some time about how to map profiles. My proposal is that we implement three profiles:

**04.07.2025 Spec updated**

| Profile | Command | Config Used | Remarks |
| :---- | :---- | :---- | :---- |
| performance | `tlp performance` *) |  **_AC** |  |
| balanced | `tlp balanced` *) | **_BAT** |  |
| power-saver | `tlp power-saver` | **_SAV**, **_BAT** |  In the first step there is not a full set of _**SAV** parameters, only the small subset ppd does atm -> https://linrunner.de/tlp/faq/ppd.html. For the rest, power-saver uses **_BAT**.|

The profile commands do *not* enter manual mode, automatic switching remains active.

*) I deliberately do not employ `tlp ac/bat`so that the behavior may be changed later if necessary.

**Comment by linrunner** [comment]
@tbodt smart-ass question (sorry) from someone with no Python experience about the dbus library used: [pydbus](https://github.com/LEW21/pydbus) has not been developed on for 7 years. Of the alternatives listed [here](https://wiki.python.org/moin/DbusExamples), only [python-sdbus](https://github.com/python-sdbus/python-sdbus) seems to be alive.

I'm unable to decipher what [tuned-ppd ](https://github.com/redhat-performance/tuned/blob/master/tuned-ppd.py) uses.

**Comment by linrunner** [comment]
Seems like GNOME 48 does not support arbitrary profiles, only performance, balanced, power-saver  ... :-(

**Comment by tbodt** [comment]
> Seems like GNOME 48 does not support arbitrary profiles, only performance, balanced, power-saver ... :-(
> 
> So how do to disable manual mode then?

Yes, KDE is the same. :( Not sure of a way. ppd doesn't have an automatic mode, the design is that the DE watches for whether you're on battery and does all the switching manually.

> [@tbodt](https://github.com/tbodt) smart-ass question (sorry) from someone with no Python experience about the D-BUS library used: [pydbus](https://github.com/LEW21/pydbus?rgh-link-date=2025-04-18T13%3A04%3A02.000Z) has not been developed on for 7 years. Of the alternatives listed [here](https://wiki.python.org/moin/DbusExamples), only [python-sdbus](https://github.com/python-sdbus/python-sdbus?rgh-link-date=2025-04-18T13%3A04%3A02.000Z) seems to be alive.
> 
> I'm unable to decipher what [tuned-ppd ](https://github.com/redhat-performance/tuned/blob/master/tuned-ppd.py?rgh-link-date=2025-04-18T13%3A04%3A02.000Z) uses.

Yes, I was disappointed by this too... tuned-ppd uses [dbus-python](https://dbus.freedesktop.org/doc/dbus-python/). Perhaps it is sensible to simply copy what it did as they have it figured out already - I figured using pydbus would be less code to write, and it is, up until you have to interface with `polkit` :P

**Comment by linrunner** [comment]
Then we'll put the auto profile topic on hold. May I assume that you agree with the rest of the proposal?

ATM in GNOME I can click on a profile but it is stuck on "Power-Saver". Is this due to the incomplete implementation of the daemon?

**Comment by tbodt** [comment]
> Then we'll put the auto profile topic on hold. May I assume that you agree with the rest of the proposal?

Looks good

> ATM in GNOME I can click on a profile but it is stuck on "Power-Saver". Is this due to the incomplete implementation of the daemon?

Yes, the setter function for the profile just prints what profile you asked for and does nothing. It should update what the getter returns, and also send a property change notification.

**Comment by linrunner** [comment]
Well, then I consider the TLP-side interface for the profiles to be defined. 

You may continue if you like. I will soon start implementing the new commands in a separate branch.

**Comment by xzl01** [comment]
Hey, I'm following your discussion. In fact, the Linux distributions we maintain use TLP as the default option for power management. It works very well, but I have some questions:
1. I want TLP to work passively, for example, TLP only switches when the user switches the power scheme, so we have to do some patches
2. In fact, tlp and tlp-dbus-daemon (implemented in the future) should be separated, and tlp-dbus-daemon can call tlp to do some power management things, so that if some users don't want to use daemon, they can also choose

**Comment by linrunner** [comment]
@tbodt Good things come to those who wait ... ;-)

It's all in the [feature/pd](https://github.com/linrunner/TLP/tree/feature/pd) branch.

You can now continue with the daemon and add the tlp calls. Note that I have slightly adapted the above [spec](https://github.com/linrunner/TLP/issues/720#issuecomment-2815324866).

**Comment by linrunner** [comment]
@xzl01 

> the Linux distributions we maintain 

Which ones exactly? Deepin?

> 1. I want TLP to work passively, for example, TLP only switches when the user switches the power scheme, so we have to do some patches

This could become a configuration parameter, for example `TLP_PASSIVE_MODE=0/1` , but it must be off by default. I won't mess with my user base.

> 2. In fact, tlp and tlp-dbus-daemon (implemented in the future) should be separated, and tlp-dbus-daemon can call tlp to do some power management things, so that if some users don't want to use daemon, they can also choose

Agreed.

**Comment by linrunner** [comment]
@ryanabx @tbodt @xzl01 the feature is now implememted in the new sub-package **tlp-pd**, see [changelog](https://github.com/linrunner/TLP/blob/main/changelog#L24).

Please test and report. Some pre-built packages are available at https://download.linrunner.de/packages/

**Comment by Mylinde** [comment]
Testing on Asus Vivobook with Ubuntu Budgie 24.04.3, Power Profiles are under Setttings > Energie present, switching is working.

<img width="1117" height="754" alt="Image" src="https://github.com/user-attachments/assets/6f3e7ea4-8de5-463c-88c0-82b7ccb3fc11" />

[stat AC](https://gist.github.com/Mylinde/a51dc7d025bdffe629fa0302914d570e)

[stat BAT](https://gist.github.com/Mylinde/cca10dbc9cefca796d1ff83bb707d700)

**Comment by linrunner** [comment]
@Mylinde Nice to hear that Budgie works too. From the screenshot I reckon its based on GNOME.

**Comment by Mylinde** [comment]
@linrunner 
Are there any differences in handling tlp settings/values to version 1.8? Have taken over the configschema and config for tlpui from version 1.8  to 1.9 to get tlpui working.

**Comment by linrunner** [comment]
@Mylinde All new or removed parameters are mentioned in the changelog.

ps. I intend to add some more before the release.

**Comment by Mylinde** [comment]
@linrunner 

Switching power profiles are not correct working. 

Performance https://gist.github.com/Mylinde/fb09c0e0a518e2a13c34eca58644e94c

Balanced https://gist.github.com/Mylinde/492a78bec047ee490b29a00f42cbb815

Power Saver https://gist.github.com/Mylinde/35ca19175771a30eeddc583bcd745cbd

Scaling driver is switching from amd-pstate-epp to amd-pstate but scaling govenor stays at performace.
Platform profile only changing from performance to quiet.

**Comment by linrunner** [comment]
@Mylinde **complete** `tlp-stat` output for each profile, please! Otherwise, no assessment is possible.

**Comment by Mylinde** [comment]
balanced BAT https://gist.github.com/Mylinde/333d8a3e9f61a9d2057af63539cab861#file-balanced_bat
saver BAT https://gist.github.com/Mylinde/527086743536777869c55b7d49bccc71#file-saver_bat
performance BAT https://gist.github.com/Mylinde/05faf0d039ffba2e1d7fdc6b361fb0c2#file-performance_bat

config https://gist.github.com/Mylinde/5c0cbb0179f756e07b071c5e6e1b0004#file-config

CPU settings now clear, scaling-driver and scaling_govenor are depending on scaling driver operation mode.
But Platform profiles are just switching if TLP_DISABLE_DEFAULTS=1 is commented out.

**Comment by linrunner** [comment]
@Mylinde so your point is: using TLP_DISABLE_DEFAULTS=1 leads to PLATFORM_PROFILE_ON_AC/BAT/SAV being disabled despite explicitly configured. Actually, there is already #786 for that. Thanks for reporting, I'll look into it of course.

**Comment by linrunner** [comment]
@Mylinde Fixed. Please re-test.

**Comment by Mylinde** [comment]
Not working profiles are not switching. PLATFORM_PROFILE_ON_AC/BAT/SAV are configured but tlp-stats shows only the PLATFORM_PROFILE_ON_SAV as configured.

https://gist.github.com/Mylinde/312b1b19ab8c75fb6ef8032eca28a16f#file-pp_dd_bat

config https://gist.github.com/Mylinde/2b77f9053c338ef50d339122416efa42#file-pp_dd_bat_config

**Comment by linrunner** [comment]
@Mylinde Oh, my! Please try again.

**Comment by Mylinde** [comment]
Working on BAT and AC, defaults disabled.

[stat](https://gist.github.com/Mylinde/6d3add842818459593b4da6b9864345e#file-fixed_pp_dd)

**Comment by linrunner** [comment]
@Mylinde thanks for testing. What exactly is your reason for using `TLP_DISABLE_DEFAULTS=1`?

**Comment by Mylinde** [comment]
Default values are usually decided by the creator and may not match the user's preferences, or they may correlate with other adjustments made by the user. I like to decide for myself which settings to change, as this makes it easier for me to identify the cause of a malfunction or poor performance without having to search for a long time.

**Comment by linrunner** [comment]
TLP 1.9.0 Beta is out. Please help with testing!
https://github.com/linrunner/TLP/issues/832

**Comment by linrunner** [comment]
The beta test is complete, and [TLP 1.9.0](https://github.com/linrunner/TLP/releases/tag/1.9.0) is now released. Have fun!  

A big thank you to all contributors, testers, and bug reporters! :+1:
